File: Python/Comic/comic.py
import requests
import os
import re
from lxml import etree


def ParseHTML(url):
    rawDoc = requests.get(url).text
    html = etree.HTML(rawDoc)
    return html


class Comic():
    def __init__(self):
        self.baseurl = "https://manhua.fzdm.com/39/"
        self.baseimgurl = "https://p5.manhuapan.com/"
        self.name = "Attack On Titan"
        self.chapters = []

    def GetChapters(self):
        html = ParseHTML(self.baseurl)
        nodes = html.xpath('//div[@id="content"]/li/a')

        for node in nodes:
            title = node.text
            url = self.baseurl + node.attrib['href']
            self.chapters.append({'title': title, 'url': url})
        return self

    def GetImgIter(self, url):
        idx = 0
        while True:
            pageurl = f"{url}index_{str(idx)}.html"
            try:
                yield {self.ExtractImg(pageurl), idx}
            except IndexError:
                return
            idx += 1

    def ExtractImg(self, pageurl):
        res = requests.get(pageurl)
        if res.status_code == 404:
            raise IndexError

        regexp = re.compile(r'(?<=mhurl=").*?(?=";)')
        rawDoc = res.text
        imgurl = self.baseimgurl + regexp.search(rawDoc).group(0)
        return imgurl

    def SaveChapter(self, chapter):
        title = chapter['title']
        url = chapter['url']
        path = f"{self.name}/{title}/"

        os.makedirs(path, exist_ok=True)

        imgurls = self.GetImgIter(url)

        for index, imgurl in imgurls:
            print(imgurl)
            # with requests.get(imgurl, stream=True) as res:
            #     with open(f"{path}{str(idx)}.jpg", "wb") as pic:
            #         for chunk in res.iter_content():
            #             pic.write(chunk)

    def Run(self):
        self.GetChapters()
        for chapter in self.chapters:
            self.SaveChapter(chapter)


c = Comic()

c.SaveChapter({'title': "test", 'url': "https://manhua.fzdm.com/39/001/"})
# c.ExtractImg("https://manhua.fzdm.com/39/001/index_100.html")

# img = c.ExtractImg("https://manhua.fzdm.com/39/001/index_1.html")

# print(img)

# for chp in c.GetChapters().chapters:
#     print(chp)

File: CSharp/Comic/Comic.cs
using System.Net.Http;
using HtmlAgilityPack;
using System.Collections.Generic;
using System;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.IO;

namespace Comic
{
    public class ComicDownloader
    {
        private readonly HttpClient _client = new();

        private string _baseUrl = "https://manhua.fzdm.com/39/";
        private string _baseImgUrl = "https://p5.manhuapan.com/";
        private string _comicName = "Attack On Titan";

        public ComicDownloader()
        {
        }

        private void CreateDirectory(string path)
        {
            if (!Directory.Exists(path))
                Directory.CreateDirectory(path);
        }


        public async Task Run()
        {
            await DownloadChapters(Chapters);
        }

        public async Task DownloadChapters(List<KeyValuePair<string, string>> chapters)
        {
            List<Task> tasks = new();
            foreach (var chapter in chapters)
                tasks.Add(DownloadChapter(chapter));

            await Task.WhenAll(tasks);
        }

        private async Task DownloadChapter(KeyValuePair<string, string> chapter)
        {
            var path = $"{_comicName}/{chapter.Key}/";

            CreateDirectory(path);

            var imgUrls = GetImageUrls(chapter.Value);

            int idx = 0;
            await foreach (var imgUrl in imgUrls)
            {
                var img = await _client.GetByteArrayAsync(imgUrl);
                File.WriteAllBytes($"{path}{idx.ToString()}.jpg", img);
                idx++;
            }
        }

        public List<KeyValuePair<string, string>> Chapters
        {
            get
            {
                var html = new HtmlWeb().Load(_baseUrl);
                var nodes = html.DocumentNode.SelectNodes("//div[@id='content']/li/a");

                List<KeyValuePair<string, string>> chapters = new();
                foreach (var node in nodes)
                {
                    string title = node.InnerText;
                    string url = _baseUrl + node.Attributes["href"].Value;
                    chapters.Add(new(title, url));
                }
                chapters.Reverse();
                return chapters;
            }
        }

        public async Task<string> ExtractImg(string url)
        {
            var res = await _client.GetAsync(url);
            if (!res.IsSuccessStatusCode)
                throw new IndexOutOfRangeException();

            Regex pattern = new(@"(?<=mhurl="").*?(?="";)");
            string rawDoc = await res.Content.ReadAsStringAsync();
            string imgUrl = _baseImgUrl + pattern.Match(rawDoc).Value;
            return imgUrl;
        }

        public async IAsyncEnumerable<string> GetImageUrls(string url)
        {
            int idx = 0;
            while (true)
            {
                string pageUrl = $"{url}index_{idx.ToString()}.html";
                string imgUrl;
                try
                {
                    imgUrl = await ExtractImg(pageUrl);
                }
                catch (IndexOutOfRangeException)
                {
                    break;
                }
                idx++;
                yield return imgUrl;
            }
        }
    }
}